File: api/AppointmentService/Models/AppointmentModel.cs
namespace AppointmentService.Models;

public class AppointmentModel
{
    public int Id { get; set; }
    public DateTime StartTime { get; set; } 
    public DateTime? EndTime { get; set; }
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string Status { get; set; } = "Pending";
    public int DoctorId { get; set; }
    public int PatientId { get; set; }
    public bool CreatedBy { get; set; } // False for Patient, True for Doctor
    public DateTime SysTimestamp { get; set; } = DateTime.Now;
    public DateTime SysCreated { get; set; } = DateTime.Now;
    
    public AppointmentModel() { }
    public AppointmentModel(int id, DateTime startTime, DateTime? endTime, string title, string? description, string status, int doctorId, int patientId, bool createdBy)
    {
        Id = id;
        StartTime = startTime;
        EndTime = endTime;
        Title = title;
        Description = description;
        Status = status;
        DoctorId = doctorId;
        PatientId = patientId;
        CreatedBy = createdBy;
        SysTimestamp = DateTime.Now; // Initialized when created
        SysCreated = DateTime.Now;  // Initialized when created
    }
    public void UpdateStatus(string newStatus)
    {
        Status = newStatus;
        SysTimestamp = DateTime.Now; // Update timestamp when a change occurs
    }
    public void UpdateDetails(string title, string? description, DateTime startTime, DateTime? endTime)
    {
        Title = title;
        Description = description;
        StartTime = startTime;
        EndTime = endTime;
        SysTimestamp = DateTime.Now;
    }
    public override string ToString()
    {
        return $"Appointment [Id={Id}, Title={Title}, Status={Status}, StartTime={StartTime}, EndTime={EndTime}, DoctorId={DoctorId}, PatientId={PatientId}]";
    }
    public bool IsOverlapping(DateTime otherStartTime, DateTime otherEndTime)
    {
        return StartTime < otherEndTime && otherStartTime < EndTime;
    }
}

File: frontend/lib/models/appointment_model.dart
import 'package:flutter/foundation.dart';

enum AppointmentStatus {
  pending,
  confirmed,
  declined,
  canceled,
  done,
}

extension AppointmentStatusExtension on AppointmentStatus {
  String toShortString() {
    return describeEnum(this).toUpperCase();
  }

  static AppointmentStatus fromString(String status) {
    switch (status.toUpperCase()) {
      case 'CONFIRMED':
        return AppointmentStatus.confirmed;
      case 'DECLINED':
        return AppointmentStatus.declined;
      case 'CANCELED':
        return AppointmentStatus.canceled;
      case 'DONE':
        return AppointmentStatus.done;
      default:
        return AppointmentStatus.pending;
    }
  }
}

class AppointmentModel {
  final int id;
  final DateTime startTime;
  final DateTime? endTime;
  final String title;
  final String? description;
  final AppointmentStatus status;
  final int doctorId;
  final int patientId;
  final bool createdBy; // False for Patient, True for Doctor
  final DateTime sysTimestamp;
  final DateTime sysCreated;

  AppointmentModel({
    required this.id,
    required this.startTime,
    this.endTime,
    required this.title,
    this.description,
    this.status = AppointmentStatus.pending,
    required this.doctorId,
    required this.patientId,
    required this.createdBy,
    required this.sysTimestamp,
    required this.sysCreated,
  });

  factory AppointmentModel.fromJson(Map<String, dynamic> json) {
    return AppointmentModel(
      id: json['id'],
      startTime: DateTime.parse(json['startTime']),
      endTime: json['endTime'] != null ? DateTime.parse(json['endTime']) : null,
      title: json['title'],
      description: json['description'],
      status: AppointmentStatusExtension.fromString(json['status']),
      doctorId: json['doctorId'],
      patientId: json['patientId'],
      createdBy: json['createdBy'],
      sysTimestamp: DateTime.parse(json['sysTimestamp']),
      sysCreated: DateTime.parse(json['sysCreated']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'startTime': startTime.toIso8601String(),
      'endTime': endTime?.toIso8601String(),
      'title': title,
      'description': description,
      'status': status.toShortString(),
      'doctorId': doctorId,
      'patientId': patientId,
      'createdBy': createdBy,
      'sysTimestamp': sysTimestamp.toIso8601String(),
      'sysCreated': sysCreated.toIso8601String(),
    };
  }

  @override
  String toString() {
    return 'AppointmentModel(id: $id, title: $title, status: $status, startTime: $startTime, endTime: $endTime, doctorId: $doctorId, patientId: $patientId)';
  }

  bool isOverlapping(DateTime otherStartTime, DateTime otherEndTime) {
    return startTime.isBefore(otherEndTime) && otherStartTime.isBefore(endTime ?? otherStartTime);
  }
}
